client/state/sites/reducer.js

```diff
@@ -98,18 +99,38 @@ export function items( state = {}, action ) {
 			return state;
 		}
 
-		case SITE_RECEIVE: {
-			const site = pick( action.site, VALID_SITE_KEYS );
-			return Object.assign( {}, state, {
-				[ site.ID ]: site
-			} );
-		}
-
+		case SITE_RECEIVE:
 		case SITES_RECEIVE:
-			return action.sites.reduce( ( memo, site ) => {
-				memo[ site.ID ] = pick( site, VALID_SITE_KEYS );
+		case SITES_UPDATE:
+			// Normalize incoming site(s) to array
+			const sites = action.site ? [ action.site ] : action.sites;
+
+			// SITES_RECEIVE occurs when we receive the entire set of user
+			// sites (replace existing state). Otherwise merge into state.
+			const initialMemo = SITES_RECEIVE === action.type ? {} : state;
+
+			return reduce( sites, ( memo, site ) => {
+				// If we're not already tracking the site upon an update, don't
+				// merge into state (TODO: Consider dropping condition once
+				// sites-list abolished)
```

Comment:
if we put the TODO on it's own line it will end up rendering better in tools which highlight and track TODOs